YNU-354: Protocol Docs (#69)
layer-3/docs: docs/protocol/terminology.mdx

Reviewer: We need to agree on clearnode usage here. As we discussed, we need to decide whether clearnode is the name of the component on a protocol level or the name of its implementation

Reviewer 2: That's why we need to standardize the terminology!

---
sidebar_position: 2
title: "Terminology"
---

import Tooltip from '@site/src/components/Tooltip';

# Terminology

## Core Concepts

**Channel**: A secure communication pathway between participants that locks funds in an on-chain smart contract while enabling off-chain state updates.

**State**: A snapshot of the channel at a point in time, including fund allocations and application-specific data.

> **Participant**: An entity (identified by an address) that is part of a channel. Typically includes a Creator (user) and a Broker (service provider).

Reviewer 3: by a wallet address

Author: Changed it to a wallet address.

> **Participant**: An entity (identified by an address) that is part of a channel. Typically includes a Creator (user) and a Broker (service provider).

Reviewer 3: Replace Broker (service provider) with Clearnode or Clearnode Operator

Author: Resolved this

> **Participant**: An entity (identified by a wallet address) that is part of a channel. Typically includes a Creator (user) and a clearnode or a Clearnode Operator.

Reviewer: I'd just say Clearnode Operator, as clearnode itself is a logical component, and we usually called this participant as broker, but now we may reconsider.

Author: <img width="824" height="581" alt="image" src="https://github.com/user-attachments/assets/62283504-6ea6-48cb-a311-49e16271b074" />
I had kept it as broker, @dimast-x  had provided a comment to change it to Clearnode or Clearnode Operator
Please let me know what should we go with

Reviewer 2: I guess @dimast-x meant "Clearnode" or "Clearnode Operator", not "Clearnode or Clearnode Operator". Personally, I am for "Clearnode", as an additional term "Operator" confuses even me. Is it a person that operates the Clearnode? Is it his personal wallet a participant then?

Author: I have kept it to Clearnode only now then.

Author: [2407a50f77b866e3cddc84621a441c69e6787515>](https://github.com/layer-3/docs/pull/69/commits/2407a50f77b866e3cddc84621a441c69e6787515)

**Adjudicator**: A smart contract that validates state transitions according to application-specific rules.

> **Broker**: An RPC message broker (also called Clear Node or Yellow Node) responsible for managing the off-chain protocol and providing services to clients.

Reviewer 3: "Clearnode" is one word. Also, don't describe that being a message broker is the main function (this description is outdated). Clearnode is a virtual ledger layer that provides a unified ledger (through Nitro RPC), and coordinates state channels (through Nitrolite), providing chain abstraction for developers and users.

Author: Addressed it throughout the protocol docs

> **Creator**: Most often a light client willing to fund a ledger account with a clearnode.

Reviewer: I'm not familiar with this term, can you describe it in more detail here in the comments?)

Reviewer: Ah, so that's a participant with zero index, I think this description should be rewritten a bit

Author: What would be the best description for a participant with zero index?

Author: How's this? @philanton 
**Creator**: The participant at index 0 in a channel who initiates channel creation. The Creator constructs the channel configuration, prepares the initial funding state, signs it (signature at position 0), and calls the on-chain `create()` function to lock their funds and establish the channel. Typically a user or light client opening a payment channel with a clearnode.

Reviewer: I think this is enough: "The participant at index 0 in a channel who initiates channel creation"
No need to go in detail as it is covered by other sections

Author: Done the same 🫡

Author: [2407a50f77b866e3cddc84621a441c69e6787515>](https://github.com/layer-3/docs/pull/69/commits/2407a50f77b866e3cddc84621a441c69e6787515)

> **Virtual Ledger Channel (VLC)**: Off-chain channels built on top of payment channels, also called App Sessions.

Reviewer 3: We dont have Virtual Ledger Channels XD. Call them straight away app sessions. and provide more description and mention that it is indented to be used by app developers

Author: Understood :) 
Did add more description as well

> **Unified Balance**: An abstraction that aggregates a user's funds across multiple blockchain networks, managed by the Broker.

Reviewer 3: by Clearnode

Author: Made changes as per this

> **Session Key**: A temporary cryptographic key delegated by a user's main wallet to enable gasless transaction signing.

Reviewer 3: remove "to enable gasless transaction signing.". Instead describe that it provides a flexible way for the user to manage security of his funds by giving specific permissions and allowances for specific apps.

Author: Added the same

> ## Identifiers

Reviewer: We can also include here `appSessionId`, `accountId` (inside unified ledger), `blockchainId` and `assetSymbol`, maybe also `userId` (identifies user after authentication into Clearnode, right now it is always user's wallet address; still not sure whether this one should be mentioned here)

Reviewer: But `wallet` or `walletAddress` I believe should be here as well

Author: Noted, would be adding that as well

Author: @philanton Would these be the right definitions:
**appSessionId**: A unique identifier for an app session, formatted as a 0x-prefixed hex string (32 bytes). Used for all subsequent operations on that specific app session.

**accountId**: An identifier for an account or app session within the unified ledger. Can be either a 0x-prefixed hex string or a wallet address.

**chainId**: A blockchain network identifier (uint64). Examples: 1 (Ethereum Mainnet), 137 (Polygon), 8453 (Base), 42161 (Arbitrum One), 10 (Optimism).

**assetSymbol**: A lowercase string identifier for a supported asset (e.g., "usdc", "eth", "weth", "usdt", "dai", "wbtc"). Asset symbols are consistent across chains.

**walletAddress**: A user's blockchain address (0x-prefixed hex string, 20 bytes) that identifies their account and owns funds. Used to identify participants in channels and app sessions.

**userId**: Identifies a user after authentication with a clearnode. Currently, this is always equivalent to the user's walletAddress.

Reviewer 2: @philanton @ihsraham , we can also mention the "userTag". Or are we going to deprecate that functionality?

Reviewer: Yep, but:
1) _chainId_ is usually `uint32`;
2) for _assetSymbol_ I woudn't mention wrapped tokens as we may use the same symbol for them, so without `wbtc` and `weth`;
3) for the _userId_ I'd only change "with a cleanode" to "to the Clearnode".

Author: Addressed

Author: [2407a50f77b866e3cddc84621a441c69e6787515>](https://github.com/layer-3/docs/pull/69/commits/2407a50f77b866e3cddc84621a441c69e6787515)

**channelId**: A unique identifier for a channel, computed as the hash of the channel configuration.

**stateHash**: A cryptographic hash of a channel state, used for signature verification.

**requestId**: A unique identifier for an RPC request, used for correlating requests and responses.

## Protocol Components

**Nitrolite**: The on-chain smart contract protocol.

**Nitro RPC**: The off-chain communication protocol.

> **Custody Contract**: The main on-chain contract implementing channel creation, joining, and closure.

Reviewer 3: Mention that Custody Contract is an implementation of the Nitrolite protocol

Author: Mentioned

:::tip Quick Reference

> These terms are used throughout the protocol specification. Bookmark this page for easy reference while reading other sections.

Reviewer 3: After updates make sure to replace these terms in all other places :)

Author: Done :)

> **Custody Contract**: The main on-chain contract implementing channel creation, joining, and closure. It is an implementation of the Nitrolite protocol.

Reviewer: I believe Custody and Adjudicator should be described in the same section

Author: So, do you mean I should make the definition be like Adjudicator or Custody Contract? Or they should be under one heading, like Core Concepts or Protocol Components ? @philanton

Reviewer: Yeah, under one heading

Author: Done

Author: [2407a50f77b866e3cddc84621a441c69e6787515>](https://github.com/layer-3/docs/pull/69/commits/2407a50f77b866e3cddc84621a441c69e6787515)

---
sidebar_position: 2

> title: "Terminology"

Reviewer 2: I am not sure whether I understand the difference between "Terminology" and "Glossary" pages. Both explain protocol concepts, but only "Glossary" terms are in tooltips, so this is confusing.

Reviewer 2: Maybe "Terminology" should in more details explain some main Glossary terms? Or "Terminology" should be renamed to "Concepts"? Not sure...

Author: Let's rename the Terminology page to "Key Concepts" or "Core Concepts"
   And add this note on the top
   
   :::tip Quick Reference
   This page provides essential protocol concepts for quick reference. For comprehensive definitions with code examples and cross-references, see the [Glossary](../glossary).
   :::
   wdyt?

Reviewer 2: I like it! @dimast-x , @philanton ?

---

import Tooltip from '@site/src/components/Tooltip';

> # Terminology

Reviewer 2: We should extract definitions into variables, and reference them here and in tooltips. This way we don't need to write the same repeated text in tooltips.
It would be a nightmare if some definition changes and we need to change it everywhere.

Author: Did a first attempt at it, will review it further

## Core Concepts

**Channel**: A secure communication pathway between participants that locks funds in an on-chain smart contract while enabling off-chain state updates.

> **State**: A snapshot of the channel at a point in time, including fund allocations and application-specific data.

Reviewer 2: ```suggestion
**State**: A snapshot of the channel at any point in time, including fund allocations and application-specific data.
```

Author: [46145fb03d3ec70aca2bc056862380c2463b59c6](https://github.com/layer-3/docs/pull/69/commits/46145fb03d3ec70aca2bc056862380c2463b59c6)

**Participant**: An entity (identified by a wallet address) that is part of a channel. Typically includes a Creator (user) and a Clearnode or Clearnode Operator.

> **Clearnode**: A virtual ledger layer that provides a unified ledger (through Nitro RPC) and coordinates state channels (through Nitrolite), providing chain abstraction for developers and users.

Reviewer 2: I think we should specify that Clearnode is a constantly running server, that provides such functionality. However, I am not sure of the best formulation.
Maybe something alike "Clearnode: a server implementing virtual ledger layer that provides ..."

Author: A server implementing a virtual ledger layer that provides a unified ledger (through Nitro RPC), coordinates state channels (through Nitrolite), and enables chain abstraction for developers and users ?

Reviewer 2: Perfect!

Author: [8c9e9266b03098b33f5e167595d73a24964b0c12](https://github.com/layer-3/docs/pull/69/commits/8c9e9266b03098b33f5e167595d73a24964b0c12)

**Creator**: The participant at index 0 in a channel who initiates channel creation. The Creator constructs the channel configuration, prepares the initial funding state, signs it (signature at position 0), and calls the on-chain `create()` function to lock their funds and establish the channel. Typically a user or light client opening a payment channel with a clearnode.

**App Sessions**: Off-chain channels built on top of payment channels, intended to be used by app developers to enable application-specific interactions and transactions without touching the blockchain.

> **Unified Balance**: An abstraction that aggregates a user's funds across multiple blockchain networks, managed by a clearnode.

Reviewer 2: ```suggestion
**Unified Balance**: An abstraction that aggregates user's funds across multiple blockchain networks, managed by a clearnode.
```

Author: [46145fb03d3ec70aca2bc056862380c2463b59c6](https://github.com/layer-3/docs/pull/69/commits/46145fb03d3ec70aca2bc056862380c2463b59c6)

**Session Key**: A temporary cryptographic key delegated by a user's main wallet that provides a flexible way for the user to manage security of their funds by giving specific permissions and allowances for specific apps.

## Identifiers

**channelId**: A unique identifier for a channel, computed as the hash of the channel configuration.

> **stateHash**: A cryptographic hash of a channel state, used for signature verification.

Reviewer 2: ```suggestion
**stateHash**: A cryptographic hash of a concatenated channelId with a channel state, used for signature verification.
```

Author: [46145fb03d3ec70aca2bc056862380c2463b59c6](https://github.com/layer-3/docs/pull/69/commits/46145fb03d3ec70aca2bc056862380c2463b59c6)

Reviewer 2: Sorry for confusion. Now it seems that we need to remove the "stateHash" definition altogether as it is not used anywhere: nor in the Custody SC, nor in the Clearnode.

**requestId**: A unique identifier for an RPC request, used for correlating requests and responses.

> **appSessionId**: A unique identifier for an app session, formatted as a 0x-prefixed hex string (32 bytes). Used for all subsequent operations on that specific app session.

Reviewer 2: > ... formatted as a 0x-prefixed hex string (32 bytes).

I think that we should either omit this or add the same for "channelId", "stateHash", "requestId" and "appSessionId".

Author: [46145fb03d3ec70aca2bc056862380c2463b59c6](https://github.com/layer-3/docs/pull/69/commits/46145fb03d3ec70aca2bc056862380c2463b59c6)

**accountId**: An identifier for an account or app session within the unified ledger. Can be either a 0x-prefixed hex string or a wallet address.

**chainId**: A blockchain network identifier (uint64). Examples: 1 (Ethereum Mainnet), 137 (Polygon), 8453 (Base), 42161 (Arbitrum One), 10 (Optimism).

**assetSymbol**: A lowercase string identifier for a supported asset (e.g., "usdc", "eth", "weth", "usdt", "dai", "wbtc"). Asset symbols are consistent across chains.

**walletAddress**: A user's blockchain address (0x-prefixed hex string, 20 bytes) that identifies their account and owns funds. Used to identify participants in channels and app sessions.

**userId**: Identifies a user after authentication with a clearnode. Currently, this is always equivalent to the user's walletAddress.

## On-Chain Contracts

> **Custody Contract**: The main on-chain contract implementing channel creation, joining, closure, and resizing. It is an implementation of the Nitrolite protocol that locks funds, tracks channel state, and coordinates with adjudicators to validate state transitions.

Reviewer 2: ```suggestion
**Custody Contract**: The main on-chain contract implementing the Nitrolite protocol. It provides the functionality to lock and unlock funds; create, close and challenge a channel; track channel state, and coordinate with adjudicators to validate state transitions on state updates.
```

Author: [46145fb03d3ec70aca2bc056862380c2463b59c6](https://github.com/layer-3/docs/pull/69/commits/46145fb03d3ec70aca2bc056862380c2463b59c6)

> **Adjudicator**: A smart contract that defines the rules for validating state transitions during disputes. The adjudicator's `adjudicate()` function is called by the Custody contract to verify whether a new state is valid based on previous states and application-specific logic. Examples include SimpleConsensus (requires both signatures) and Remittance (only sender must sign).

Reviewer 2: ```suggestion
**Adjudicator**: A smart contract that defines the rules for validating state transitions during all channel operations. The adjudicator's `adjudicate(...)` function is called by the Custody contract to verify whether a new state is valid based on previous states and application-specific logic. Examples include SimpleConsensus (requires both signatures) and Remittance (only sender must sign).
```

Author: [46145fb03d3ec70aca2bc056862380c2463b59c6](https://github.com/layer-3/docs/pull/69/commits/46145fb03d3ec70aca2bc056862380c2463b59c6)
